File: rust/src/368.rs
//368. Largest Divisible Subset

impl Solution {
    pub fn largest_divisible_subset(mut nums: Vec<i32>) -> Vec<i32> {
        if nums.is_empty() {
            return vec![];
        }

        nums.sort();
        let n = nums.len();
        let mut dp = vec![1; n];
        let mut prev = vec![-1; n];
        let mut max_idx = 0;

        for i in 1..n {
            for j in 0..i {
                if nums[i] % nums[j] == 0 && dp[j] + 1 > dp[i] {
                    dp[i] = dp[j] + 1;
                    prev[i] = j as i32;
                }
            }
            if dp[i] > dp[max_idx] {
                max_idx = i;
            }
        }

        let mut result = vec![];
        let mut i = max_idx as i32;
        while i >= 0 {
            result.push(nums[i as usize]);
            i = prev[i as usize];
        }

        result.reverse();
        result
    }
}
//sub: https://leetcode.com/problems/largest-divisible-subset/submissions/1598125911/?envType=daily-question&envId=2025-04-06
//sol: https://leetcode.com/problems/largest-divisible-subset/solutions/6620526/368-largest-divisible-subset-by-nalindal-m15d/

File: ts/src/368.ts
//1733. Minimum Number of People to Teach
function minimumTeachings(
  n: number,
  languages: number[][],
  friendships: number[][],
): number {
  const m = languages.length; // number of users

  // Step 1: build language sets for each user (1-indexed)
  const langSet: Set<number>[] = new Array(m + 1);
  for (let i = 0; i <= m; i++) {
    langSet[i] = new Set();
  }
  for (let i = 0; i < m; i++) {
    for (const l of languages[i]) {
      langSet[i + 1].add(l);
    }
  }

  // Step 2: find candidates (users in friendships that can't communicate)
  const candidates = new Set<number>();
  for (const [u, v] of friendships) {
    let canCommunicate = false;
    for (const l of langSet[u]) {
      if (langSet[v].has(l)) {
        canCommunicate = true;
        break;
      }
    }
    if (!canCommunicate) {
      candidates.add(u);
      candidates.add(v);
    }
  }

  if (candidates.size === 0) return 0;

  // Step 3: try teaching each language
  let ans = Number.MAX_SAFE_INTEGER;
  for (let lang = 1; lang <= n; lang++) {
    let cnt = 0;
    for (const user of candidates) {
      if (!langSet[user].has(lang)) {
        cnt++;
      }
    }
    ans = Math.min(ans, cnt);
  }

  return ans;
}

//sub: https://leetcode.com/problems/minimum-number-of-people-to-teach/submissions/1765651117/?envType=daily-question&envId=2025-09-10
//sol: https://leetcode.com/problems/minimum-number-of-people-to-teach/solutions/7174219/1733-minimum-number-of-people-to-teach-b-gudr/
